fuse/file.go

```diff
@@ -3,103 +3,370 @@ package fuse
 import (
 	"context"
 	"io"
+	"os"
+	"sync"
 
 	"bazil.org/fuse"
 	"bazil.org/fuse/fs"
+	"bazil.org/fuse/fuseutil"
 	"github.com/puppetlabs/wash/activity"
 	"github.com/puppetlabs/wash/plugin"
 )
 
 // ==== FUSE file Interface ====
 
+// file's implementation differs primarily on two axes: whether `Size` is set, and whether we're
+// currently writing.
+// - a *file-like* entry declares a `Size` in its `Attributes`
+//   - reads and writes are symmetric; the kernel page cache will be used, and the file size
+//     represents the current local state of the file
+//   - when writing, `data` represents the local state of the file; changes to its size will be
+//     reflected in `data` and reads will be served from it. The file's size will be `len(data)`.
+//     Unchanged sections will be filled from `plugin.Read` as needed.
+//   - when not writing, data is read directly from `plugin.Read`
+// - a *non-file-like* entry has `Size` unset
+//   - read always pulls from `plugin.Read` and writes are buffered independently
+//   - `data` stores only the data to be written, and is not initialized from `plugin.Read`
+//   - the file's size will be reported as its readable size; it will not reflect calls to `write`
+//
+// Note that writes only result in calling `plugin.Write` when a file handle is closed. Writing
```

Comment:
No. This is because FUSE's semantics are confusing. When the system call `close` is called on a file handle, FUSE calls the Flush callback. When the last copy of a file handle is closed, FUSE calls the Release callback. See the description at https://libfuse.github.io/doxygen/structfuse__operations.html#ad4ec9c309072a92dd82ddb20efa4ab14. Would a link to that help?

Reply:
Yes, that would help.

Reply:
Done.